### ISSUE TITLE
Relational model's attributes, when returned in a collection not properly being mapped when type `json`

### ISSUE THREAD

**Issue by DWC01** [scoutforpets/jsonapi-mapper#101]
Hey There, 

So I'm noticing an issue when the jsonapi-mapper returns a collection, and there are nested models in the collection (`hasOne` relationship in Bookshelf), and one of the attributes in the nested model is type `json` (db is Postgresql). In this case all of the nested models get populated with the same data as the last nested model.. 

My two models:
1.) company
2.) company_image

In Bookshelf I have set the company -> company_image relationship as `hasOne`..

Please see below the raw Bookshelf response for the `/companies` query..

```json
// 20180127161646
// http://localhost/companies

[
  {
    "id": 1,
    "name": "MSR",
    "description": "In 1969, Seattle engineer and lifelong mountaineer Larry Penberthy formed Mountain Safety Research as a one-man crusade dedicated to improving the safety of climbing equipment. The fuel behind Larrys passionate fire was a simple belief that still drives our team today",
    "company_type": "Larger Brand",
    "logo_url": null,
    "country": "USA",
    "website_url": "https://www.msrgear.com/",
    "updated_at": null,
    "created_at": "2018-01-27T07:31:19.386Z",
    "company_image": {
      "id": 1,
      "name": "msr-logo.jpg",
      "url": "https://s3-us-west-1.amazonaws.com/pack-planner-backend/development/temp/uploads/images/36f0bed0-0334-11e8-9a7c-293f05681529/msr-logo.jpg",
      "metadata": {
        "name": "msr-logo.jpg",
        "mimetype": "image/jpeg",
        "encoding": "7bit"
      },
      "links": {
        "logo": "https://s3-us-west-1.amazonaws.com/pack-planner-backend/development/uploads/images/1/logo/msr-logo.jpg",
        "list": "https://s3-us-west-1.amazonaws.com/pack-planner-backend/development/uploads/images/1/list/msr-logo.jpg"
      },
      "company_id": 1,
      "created_at": "2018-01-27T07:31:19.453Z",
      "updated_at": null
    }
  },
  {
    "id": 2,
    "name": "Gregory",
    "description": "At age 14 Wayne Gregory designed his first backpack as part of a Boy Scout project. Shortly after this experience he met Andy Drollinger, the founder of San Diego California Adventure 16, at a boy scout camporee. Andy was impressed with Waynes design and over the next few years Wayne could be found hanging around the Adventure 16 factory, buying raw materials and tinkering with a variety of backpack ideas. He soon became the young companies second employee.",
    "company_type": "Larger Brand",
    "logo_url": null,
    "country": "USA",
    "website_url": "https://www.gregorypacks.com/",
    "updated_at": null,
    "created_at": "2018-01-27T07:31:19.407Z",
    "company_image": {
      "id": 2,
      "name": "gregory-logo.png",
      "url": "https://s3-us-west-1.amazonaws.com/pack-planner-backend/development/temp/uploads/images/93f9fe60-03ad-11e8-9a7c-293f05681529/gregory-logo.png",
      "metadata": {
        "name": "gregory-logo.png",
        "mimetype": "image/png",
        "encoding": "7bit"
      },
      "links": {
        "logo": "https://s3-us-west-1.amazonaws.com/pack-planner-backend/development/uploads/images/2/logo/gregory-logo.png",
        "list": "https://s3-us-west-1.amazonaws.com/pack-planner-backend/development/uploads/images/2/list/gregory-logo.png"
      },
      "company_id": 2,
      "created_at": "2018-01-27T07:31:19.484Z",
      "updated_at": null
    }
  }
]
```

You can see above that for the `links` and `metadata` attributes, in the company_image model are distinctly different for each company. However, when I serialize this same data with jsonapi-mapper, I get the below output..

Below you will notice that in the `included` attribute for company 1, that the `metadata` and `links` fields (both db field type `json`) in the company_image are returned with same data as the second company_image. It seems that all other fields, that were not type `json` were properly mapped in this collection..

```json
/ 20180127155816
// http://localhost/companies

{
  "data": [
    {
      "type": "companies",
      "id": "1",
      "attributes": {
        "name": "MSR",
        "description": "In 1969, Seattle engineer and lifelong mountaineer Larry Penberthy formed Mountain Safety Research as a one-man crusade dedicated to improving the safety of climbing equipment. The fuel behind Larrys passionate fire was a simple belief that still drives our team today",
        "company_type": "Larger Brand",
        "logo_url": null,
        "country": "USA",
        "website_url": "https://www.msrgear.com/",
        "updated_at": null,
        "created_at": "2018-01-27T07:31:19.386Z"
      },
      "relationships": {
        "company_image": {
          "data": {
            "type": "company_images",
            "id": "1"
          }
        }
      }
    },
    {
      "type": "companies",
      "id": "2",
      "attributes": {
        "name": "Gregory",
        "description": "At age 14 Wayne Gregory designed his first backpack as part of a Boy Scout project. Shortly after this experience he met Andy Drollinger, the founder of San Diego California Adventure 16, at a boy scout camporee. Andy was impressed with Waynes design and over the next few years Wayne could be found hanging around the Adventure 16 factory, buying raw materials and tinkering with a variety of backpack ideas. He soon became the young companies second employee.",
        "company_type": "Larger Brand",
        "logo_url": null,
        "country": "USA",
        "website_url": "https://www.gregorypacks.com/",
        "updated_at": null,
        "created_at": "2018-01-27T07:31:19.407Z"
      },
      "relationships": {
        "company_image": {
          "data": {
            "type": "company_images",
            "id": "2"
          }
        }
      }
    }
  ],
  "included": [
    {
      "type": "company_images",
      "id": "1",
      "attributes": {
        "name": "msr-logo.jpg",
        "url": "https://s3-us-west-1.amazonaws.com/pack-planner-backend/development/temp/uploads/images/36f0bed0-0334-11e8-9a7c-293f05681529/msr-logo.jpg",
        "metadata": {
          "name": "gregory-logo.png",
          "mimetype": "image/png",
          "encoding": "7bit"
        },
        "links": {
          "logo": "https://s3-us-west-1.amazonaws.com/pack-planner-backend/development/uploads/images/2/logo/gregory-logo.png",
          "list": "https://s3-us-west-1.amazonaws.com/pack-planner-backend/development/uploads/images/2/list/gregory-logo.png"
        },
        "company_id": 1,
        "created_at": "2018-01-27T07:31:19.453Z",
        "updated_at": null
      }
    },
    {
      "type": "company_images",
      "id": "2",
      "attributes": {
        "name": "gregory-logo.png",
        "url": "https://s3-us-west-1.amazonaws.com/pack-planner-backend/development/temp/uploads/images/93f9fe60-03ad-11e8-9a7c-293f05681529/gregory-logo.png",
        "metadata": {
          "name": "gregory-logo.png",
          "mimetype": "image/png",
          "encoding": "7bit"
        },
        "links": {
          "logo": "https://s3-us-west-1.amazonaws.com/pack-planner-backend/development/uploads/images/2/logo/gregory-logo.png",
          "list": "https://s3-us-west-1.amazonaws.com/pack-planner-backend/development/uploads/images/2/list/gregory-logo.png"
        },
        "company_id": 2,
        "created_at": "2018-01-27T07:31:19.484Z",
        "updated_at": null
      }
    }
  ]
}
```

For my actual code I'm setting this up in the below fashion:

```javascript
const mapper = new Mapper.Bookshelf('http://packplanner.io');

return mapper.map(models, 'company',  {
      relations: true,
      enableLinks: false
});
```

Is this a known issue, or something that I can add some type of custom serialization on my end to avoid? Any feedback would be greatly appreciated, thanks.

**Comment by DWC01** [comment]
To add a little more context, below is the raw Bookshelf response for the same `/companies` query..
```json
// 20180127161646
// http://localhost/companies

[
  {
    "id": 1,
    "name": "MSR",
    "description": "In 1969, Seattle engineer and lifelong mountaineer Larry Penberthy formed Mountain Safety Research as a one-man crusade dedicated to improving the safety of climbing equipment. The fuel behind Larrys passionate fire was a simple belief that still drives our team today",
    "company_type": "Larger Brand",
    "logo_url": null,
    "country": "USA",
    "website_url": "https://www.msrgear.com/",
    "updated_at": null,
    "created_at": "2018-01-27T07:31:19.386Z",
    "company_image": {
      "id": 1,
      "name": "msr-logo.jpg",
      "url": "https://s3-us-west-1.amazonaws.com/pack-planner-backend/development/temp/uploads/images/36f0bed0-0334-11e8-9a7c-293f05681529/msr-logo.jpg",
      "metadata": {
        "name": "msr-logo.jpg",
        "mimetype": "image/jpeg",
        "encoding": "7bit"
      },
      "links": {
        "logo": "https://s3-us-west-1.amazonaws.com/pack-planner-backend/development/uploads/images/1/logo/msr-logo.jpg",
        "list": "https://s3-us-west-1.amazonaws.com/pack-planner-backend/development/uploads/images/1/list/msr-logo.jpg"
      },
      "company_id": 1,
      "created_at": "2018-01-27T07:31:19.453Z",
      "updated_at": null
    }
  },
  {
    "id": 2,
    "name": "Gregory",
    "description": "At age 14 Wayne Gregory designed his first backpack as part of a Boy Scout project. Shortly after this experience he met Andy Drollinger, the founder of San Diego California Adventure 16, at a boy scout camporee. Andy was impressed with Waynes design and over the next few years Wayne could be found hanging around the Adventure 16 factory, buying raw materials and tinkering with a variety of backpack ideas. He soon became the young companies second employee.",
    "company_type": "Larger Brand",
    "logo_url": null,
    "country": "USA",
    "website_url": "https://www.gregorypacks.com/",
    "updated_at": null,
    "created_at": "2018-01-27T07:31:19.407Z",
    "company_image": {
      "id": 2,
      "name": "gregory-logo.png",
      "url": "https://s3-us-west-1.amazonaws.com/pack-planner-backend/development/temp/uploads/images/93f9fe60-03ad-11e8-9a7c-293f05681529/gregory-logo.png",
      "metadata": {
        "name": "gregory-logo.png",
        "mimetype": "image/png",
        "encoding": "7bit"
      },
      "links": {
        "logo": "https://s3-us-west-1.amazonaws.com/pack-planner-backend/development/uploads/images/2/logo/gregory-logo.png",
        "list": "https://s3-us-west-1.amazonaws.com/pack-planner-backend/development/uploads/images/2/list/gregory-logo.png"
      },
      "company_id": 2,
      "created_at": "2018-01-27T07:31:19.484Z",
      "updated_at": null
    }
  }
]
```

You can see above that the `links` and `metadata` attributes are distinctly different for each company-image. However, when I serialize this same data with jsonapi-mapper, I get the output pasted in the above comment.. 

For my actual code I'm setting this up in the below fashion:

```javascript
const mapper = new Mapper.Bookshelf('http://packplanner.io');

return mapper.map(models, 'company',  {
      relations: true,
      enableLinks: false
});
```

**Comment by chamini2** [comment]
This is **not** a known issue. What version of jsonapi-mapper are you using?

**Comment by chamini2** [comment]
In my experience, json convertion does work. We are using the 1.0.0-beta.15.

**Comment by chamini2** [comment]
An equivalent test is working for me fine. We are on version 1.0.0-beta.15.

**Comment by flmg** [comment]
I have the same issue on 1.0.0-beta15. I have a `users` table with a `roles` table containing a `user_id` column and an array of roles `roles`. When I print the models before the mapping:

        data.models.forEach(m => {
            console.log('relation', m.relations.roles.attributes);
        });

        data = mapper.map(data, 'user', {
            pagination: data.pagination,
            meta: data.pagination
        });

It outputs as expected:
    
    relation { user_id: '4', roles: [ 'admin' ], test: '{admin}' }
    relation { user_id: '104', roles: [ 'manager' ], test: '{manager}' }

But after the mapping I receive this:

    "included": [
        {
            "type": "roles",
            "id": "4",
            "attributes": {
                "roles": [
                    "manager"
                ],
                "test": "{admin}"
            },
            ...
        },
        {
            "type": "roles",
            "id": "104",
            "attributes": {
                "roles": [
                    "manager"
                ],
                "test": "{manager}"
            },
            ...
        }
    ]

I added the `test` column (a text equivalent of the `roles` column to check, the problem only seems to happen on the array column (I am using a postgresql database by the way).

**Comment by chamini2** [comment]
Could you show the `CREATE TABLE` statement to check the types of the columns?

**Comment by flmg** [comment]
@chamini2 

    create table user_roles (
        user_id    bigint        references users(id) on delete cascade constraint user_roles_pkey primary key,
        roles      varchar(32)[]
    );

**Comment by chamini2** [comment]
Ok, seems to be fixed in https://github.com/scoutforpets/jsonapi-mapper/pull/102. Could you give it a try?

**Comment by flmg** [comment]
Tested, it works 🎉 Thanks!

**Comment by chamini2** [comment]
No problem!

**Comment by chamini2** [comment]
Released 1.0.0-beta.16

**Comment by dc-devs** [comment]
@chamini2  Sorry for delay here as its been a pretty busy on my end, confirming that this fix works great (with column type `json` as well). Thanks for the quick turnaround!
